File: codejam/python/2008/1B/1B_Mousetrap.py
import sys
from collections import deque

## Input crap
infile = sys.stdin
intokens = deque()
def getTokens(): 
    while not intokens:
        for c in infile.readline().rstrip().split() :
            intokens.append(c)    
def gs(): getTokens(); return intokens.popleft()
def gi(): return int(gs())
def gf(): return float(gs())
def gbs(): return [c for c in gs()]
def gis(n): return [gi() for i in range(n)]
def ia(m): return [0] * m
def iai(m,v): return [v] * m
def twodi(n,m,v): return [iai(m,v) for i in range(n)]
def fill2(m) : r = gis(2*m); return r[0::2],r[1::2]
def fill3(m) : r = gis(3*m); return r[0::3],r[1::3],r[2::3]
def fill4(m) : r = gis(4*m); return r[0::4],r[1::4],r[2::4],r[3::4]
MOD = 998244353
##MOD = 1000000007

class ds :
    def __init__(self,n=1) :
        self.n = n
        self.sz = 1
        while self.sz < n : self.sz *= 2
        self.a = [0] * (2*self.sz)
        for i in range(self.sz,self.sz+n) : self.a[i] = 1
        for i in range(self.sz-1,-1,-1) : self.a[i] = self.a[2*i]+self.a[2*i+1]
    def _query(self,idx,n) :
        if idx >= self.sz : return idx-self.sz
        if self.a[2*idx] >= n : return self._query(2*idx,n)
        return self._query(2*idx+1,n-self.a[2*idx])
    def query(self,n) : return self._query(1,n)
    def use(self,n) :
        idx= self.sz+n
        while idx > 0 : self.a[idx] -= 1; idx >>= 1

def main() :
    if len(sys.argv) > 1 : global infile; infile = open(sys.argv[1],'rt')
    ## PROGRAM STARTS HERE
    T = gi()
    for tt in range(1,T+1) :
        K = gi(); n = gi(); d = gis(n)
        unused = K; curs = 0; deck = ds(K)
        ansarr = [0] * K
        for i in range(1,K+1) :
            v = (curs+i) % unused
            if v == 0 : v = unused
            x = deck.query(v)
            deck.use(x)
            ansarr[x] = i
            curs = v-1
            unused -= 1
        ans = [ ansarr[dd-1] for dd in d ]
        ansstr = " ".join([str(x) for x in ans])
        print(f"Case #{tt}: {ansstr}")

if __name__ == "__main__" :
    main()

File: codejam/cpp/2008/1B/1B_Mousetrap.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<ll> vi;
typedef pair<ll,ll> pi;
#define FOR(i,a) for (ll i = 0; i < (a); i++)
#define len(x) (ll) x.size()
const ll INF = 1LL << 62;
const ll MOD = 1000000007;
//const ll MOD = 998244353;
const double PI = 4*atan(double(1.0));

struct ds {
    private :
        ll _n,sz;
        vi arr;
    public :
        ds() : _n(0),sz(0) {}
        ds(ll n) {
            _n=n;
            sz = 1; while(sz<n) { sz <<= 1; }
            arr.resize(2*sz,0);
            for (ll i = sz; i < sz+n; i++ ) { arr[i] = 1; }
            for (ll i = sz-1; i >= 1; i-- ) { arr[i] = arr[2*i] + arr[2*i+1]; }
        }
        ll _query(ll idx, ll n) {
            if (idx >= sz) return idx-sz;
            if (arr[2*idx] >= n) return _query(2*idx,n);
            return _query(2*idx+1,n-arr[2*idx]);
        }
        ll query(ll n) { return _query(1,n); }
        void use(ll n) {
            ll idx = sz+n;
            while (idx > 0) { arr[idx] -= 1; idx >>= 1; }
        }
};

int main (int argc, char **argv) {
    if (argc > 1) { freopen(argv[1],"r",stdin); }
    ios_base::sync_with_stdio(false);cin.tie(0);
    // PROGRAM STARTS HERE
    ll T; cin >> T;
    for (ll tt=1;tt<=T;tt++) {
        ll K,n; cin >> K >> n;
        vi d(n); FOR(i,n) { cin >> d[i]; }
        ll unused(K), curs(0);
        auto deck = ds(K);
        vi ansarr(K,0);
        for (ll i = 1; i <= K; i++ ) {
            ll v = (curs+i) % unused;
            if (v == 0) { v = unused; }
            auto x = deck.query(v);
            deck.use(x);
            ansarr[x] = i;
            curs = v-1; 
            unused -= 1;
        }
        vi ans(n); FOR(i,n) { ans[i] = ansarr[d[i]-1]; }
        cout << "Case #" << tt << ": " << ans[0];
        for (ll i = 1; i < n; i++) { cout << " " << ans[i]; }
        cout << '\n';
    }
}

